client/db/src/lib.rs

```diff
@@ -1399,6 +1498,49 @@ impl<Block: BlockT> Backend<Block> {
 
 		Ok(())
 	}
+
+	fn prune_blocks(
+		&self,
+		transaction: &mut Transaction<DbHash>,
+		finalized: NumberFor<Block>,
+	) -> ClientResult<()> {
+		if let KeepBlocks::Some(keep_blocks) = self.keep_blocks {
+			// Always keep the last finalized block
+			let keep = std::cmp::max(keep_blocks, 1);
+			if finalized >= keep.into() {
+				let number = finalized.saturating_sub(keep.into());
+				match read_db(&*self.storage.db, columns::KEY_LOOKUP, columns::BODY, BlockId::<Block>::number(number))? {
+					Some(body) => {
+						utils::remove_db(
+							transaction,
+							&*self.storage.db,
+							columns::KEY_LOOKUP,
+							columns::BODY,
+							BlockId::<Block>::number(number)
+						)?;
+						debug!(target: "db", "Removing block #{}", number);
+						match self.transaction_storage {
+							TransactionStorageMode::BlockBody => {},
+							TransactionStorageMode::StorageChain => {
+								match Vec::<Block::Hash>::decode(&mut &body[..]) {
+									Ok(hashes) => {
+										for h in hashes {
+											transaction.remove(columns::TRANSACTION, h.as_ref());
+										}
+									}
+									Err(err) => return Err(sp_blockchain::Error::Backend(
+										format!("Error decoding body list: {}", err)
+									)),
+								}
+							}
+						}
```

Comment:
```suggestion
						if let TransactionStorageMode::StorageChain = self.transaction_storage {
							match Vec::<Block::Hash>::decode(&mut &body[..]) {
								Ok(hashes) => {
									for h in hashes {
										transaction.remove(columns::TRANSACTION, h.as_ref());
									}
								}
								Err(err) => return Err(sp_blockchain::Error::Backend(
									format!("Error decoding body list: {}", err)
								)),
							}
						}
```

Reply:
Would really prefer to keep `match` here for the same reason `match` is required to be exhaustive in rust. Code update will be required In case a new variant is added to the enum.